net: lib: coap: Service support

pdgendt opened the pull request:
This PR introduces CoAP services (similar to HTTP services) and a CoAP server implementation.

~Note: the [CoAP Server sample](https://docs.zephyrproject.org/latest/samples/net/sockets/coap_server/README.html) was copied as everything remains backwards compatible.~ If this isn't desired it can be removed in favor of CoAP services.

pdgendt committed:
net: lib: coap: Add service support

Add CoAP services and server as a subsystem implementation.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

pdgendt committed:
tests: net: lib: Add CoAP service tests

Add a test for CoAP macro usage.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

pdgendt committed:
net: lib: coap: Add resources length based variants

Add length variant for the well known core resource and parsing.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

pdgendt committed:
net: lib: coap: Add coap_uri_path_match

Add URI path matching function to public CoAP API.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

pdgendt committed:
net: lib: coap: Add coap_packet_is_request

Add function to check if CoAP packet is a request to the public API.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

pdgendt committed:
net: lib: coap: Add CoAP server shell

Add shell commands that allow to start/stop CoAP services.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

pdgendt committed:
samples: net: sockets: Add CoAP service example

Replaced the CoAP server example with CoAP services.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

pdgendt committed:
net: lib: coap: Add coap_find_observer_by_token

Add a CoAP helper function to find a matching observer by token.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

pdgendt committed:
net: lib: coap: coap_remove_observer result type

Change coap_remove_observer to return the result of removing the
observer if found.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

pdgendt committed:
net: lib: coap: Add support for observer event callbacks

This commit adds the option to register an event handler to CoAP
resources when observers are added/removed.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

pdgendt committed:
include: coap: Fix typo

Fixed a typo to include DOXYGEN generated documentation.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

jukkar commented on samples/net/sockets/coap_service/src/main.c:
```diff
@@ -0,0 +1,85 @@
+
+	iface = net_if_get_default();
+	if (!iface) {
+		LOG_ERR("Could not get te default interface");
```
typo, s/te/the/

jukkar commented on samples/net/sockets/coap_service/src/test.c:
```diff
@@ -0,0 +1,262 @@
+
+	code = coap_header_get_code(request);
+	type = coap_header_get_type(request);
+	id = coap_header_get_id(request);
+	tkl = coap_header_get_token(request, token);
+
+	LOG_INF("*******");
+	LOG_INF("type: %u code %u id %u", type, code, id);
+	LOG_INF("*******");
+
+	payload = coap_packet_get_payload(request, &payload_len);
+	if (payload) {
+		net_hexdump("POST Payload", payload, payload_len);
+	}
+
+	if (type == COAP_TYPE_CON) {
+		type = COAP_TYPE_ACK;
+	} else {
+		type = COAP_TYPE_NON_CON;
+	}
+
+	r = coap_packet_init(&response, data, sizeof(data),
+			     COAP_VERSION_1, type, tkl, token,
+			     COAP_RESPONSE_CODE_CREATED, id);
+	if (r < 0) {
+		return r;
+	}
+
+	for (p = location_path; *p; p++) {
+		r = coap_packet_append_option(&response,
+					      COAP_OPTION_LOCATION_PATH,
+					      *p, strlen(*p));
+		if (r < 0) {
+			return r;
+		}
+	}
+
+	r = coap_resource_send(resource, &response, addr, addr_len);
+
+	return r;
```
Just pondering here, the functions look pretty similar to me. Would it be possible to refactor them to share the common code?

pdgendt replied:
This code was copied from the current `coap_server` sample with minimal changes, to prove the backwards compatibility.

jukkar commented on include/zephyr/net/coap_service.h:
```diff
@@ -0,0 +1,249 @@
+ * @param[inout] _port Pointer to port associated with the service.
+ * @param _flags Configuration flags @see @ref COAP_SERVICE_FLAGS.
+ */
+#define COAP_SERVICE_DEFINE(_name, _host, _port, _flags)					\
```
Does this mean that all the services need to be defined at built time?

pdgendt replied:
That's the idea, yes. They can be started/stopped at runtime, but a single server thread is used for all services.

pdgendt replied:
Both services and resources are placed in predefined linker sections.

pdgendt replied:
Right, I'll add a note to the function brief, and add a check/assert.

rlubos commented on subsys/net/lib/coap/coap_server.c:
```diff
@@ -0,0 +1,749 @@
+					      options, opt_num, &client_addr, client_addr_len);
+
+		/* Shortcut for replying a code without a body */
+		if (ret > 0  && ret <= UINT8_MAX) {
```
What's the meaning of `UINT8_MAX` in this case? Plus, I'm a bit confused regarding the documentation of `coap_handle_request_len()`, it only mentions that `0` is returned in case of success.

pdgendt replied:
Yes, this "feature" should be explained a bit more in the documentation.

Both `coap_handle_request` and `coap_handle_request_len` return the value of the handler called in the resource. I think it's very nice to have the option to send a response code (either success or error), without the need to build an acknowledge buffer in each handler. An example:

```c
static const struct gpio_dt_spec led = GPIO_DT_SPEC_GET_OR(DT_ALIAS(led0), gpios, {0});

static int led_put(struct coap_resource *resource, struct coap_packet *request,
		   struct sockaddr *addr, socklen_t addr_len)
{
	const uint8_t *payload;
	uint16_t payload_len;

	if (!gpio_is_ready_dt(&led)) {
		return COAP_RESPONSE_CODE_INTERNAL_ERROR;
	}

	payload = coap_packet_get_payload(request, &payload_len);
	if (payload_len != 1) {
		return COAP_RESPONSE_CODE_BAD_REQUEST;
	}

	if (gpio_pin_set_dt(&led, payload[0]) < 0) {
		return COAP_RESPONSE_CODE_INTERNAL_ERROR;
	}

	return COAP_RESPONSE_CODE_CHANGED;
}
```

With the shortcut feature, a response will be sent from the underlying server code.

pdgendt replied:
Should this be turned into a `Kconfig` option so users are aware? @rlubos

rlubos replied:
I don't think there's a need for a Kconfig option here, personally I don't have an issue with this feature, as it seems reasonable, it'd just make sense to document it somewhere. Maybe we should elaborate on the expected method handlers behavior in `COAP_RESOURCE_DEFINE()` documentation? That if they don't need to include payload in the response, they can just return the response code to reply with.

I think it'd also make sense to extend the existing CoAP documentation we have (https://github.com/zephyrproject-rtos/zephyr/blob/main/doc/connectivity/networking/api/coap.rst), or add a separate page (like for CoAP client helper lib), to raise awareness of this new CoAP service functionality (which btw seems pretty nice and helpful!).

rlubos commented on include/zephyr/net/coap_service.h:
```diff
@@ -0,0 +1,255 @@
+ * @param addr Peer address
+ * @retval 1 in case an observer was removed.
+ * @retval 0 in case no observer was removed.
+ * @return negative value in case of an error.
```
Mixing `retval` with `return` does not render well in API docs, we end up with two "Return" sections:
https://builds.zephyrproject.io/zephyr/pr/64265/docs/doxygen/html/group__coap__service.html#ga8d9ab0bf6b1ea15408f1c80c45aae16b

Plus, I wonder about the return values, wouldn't it make sense to return `0` in case of success (observer removed) and treat missing observer as an error condition `-ENOENT`?

pdgendt replied:
Updated the documentation, and applied te suggestion.

rlubos commented on samples/net/sockets/coap_service/README.rst:
```diff
@@ -0,0 +1,57 @@
+Overview
+********
+
+This sample is a simple CoAP server showing how to expose a simple resource.
```
So the question is - shall this new sample be considered a replacement for the old `coap_server` sample? I don't think there's a need to have two samples exposing similar functionality.

pdgendt replied:
Replaced the existing sample using the new services.

kartben commented on samples/net/sockets/coap_server/README.rst:
```diff
@@ -1,8 +1,8 @@
    :relevant-api: coap udp
 
-   Use the CoAP library to implement a server that exposes CoAP resources.
+   Use the CoAP server subsystem that exposes CoAP resources.
```
better wording below?

```suggestion
   Use the CoAP server subsystem to register CoAP resources.
```

kartben replied:
It feels like the rest of the README could be slightly reworded too and provide a brief explanation of how it now uses the CoAP service?

pdgendt commented:
Rebased, and resolved merge conflict

kartben commented on samples/net/sockets/coap_server/README.rst:
```diff
@@ -1,8 +1,8 @@
 .. zephyr:code-sample:: coap-server
-   :name: CoAP server
+   :name: CoAP service
    :relevant-api: coap udp
```
```suggestion
   :relevant-api: coap coap_service udp
```

kartben commented on include/zephyr/net/coap_service.h:
```diff
@@ -0,0 +1,284 @@
+
+/**
+ * @brief CoAP Service API
+ * @defgroup coap_service CoAP service API
```
I would add this to the API Reference section of the `connectivity/networking/api/coap`  page

pdgendt replied:
So, not creating a new group? It does depend on an additional `Kconfig` setting besides `CONFIG_COAP` though.

kartben replied:
I think a new group is fine, I was simply recommending the following so that it actually shows up in the Sphinx docs:

```diff
diff --git a/doc/connectivity/networking/api/coap.rst b/doc/connectivity/networking/api/coap.rst
index 4f0654a57d..eae20b1f36 100644
--- a/doc/connectivity/networking/api/coap.rst
+++ b/doc/connectivity/networking/api/coap.rst
@@ -204,3 +204,4 @@ API Reference
 *************

 .. doxygengroup:: coap
+.. doxygengroup:: coap_service
```

On the other hand, it looks like coap_client has its own documentation page, so I will let @rlubos and @jukkar comment on what they think is best.

rlubos replied:
Sorry for the delay, I somehow missed the comment... I think that more elaborate documetnation (like we have for CoAP client) would be ideal, but I don't want to block the PR with this. At the very least we should however reference the APIs in the CoAP doc page, as @kartben suggests.

pdgendt replied:
I will add a documentation page with examples, either to this, or a follow-up PR. 
Currently CoAP secure `coaps://` is higher on my priority list, which can also be added to this PR or the next

pdgendt replied:
The secure part will take more time as DTLS servers currently only support 1 client, I've added a documentation page.

kartben commented on include/zephyr/net/coap_link_format.h:
```diff
@@ -34,6 +34,12 @@ int coap_well_known_core_get(struct coap_resource *resource,
 			     struct coap_packet *response,
 			     uint8_t *data, uint16_t len);
 
+int coap_well_known_core_get_len(struct coap_resource *resources,
```
missing docstring (or maybe it's not meant to be a public API? docstring also missing for coap_well_known_core_get() just above.

kartben commented on include/zephyr/net/coap.h:
```diff
@@ -238,6 +238,25 @@ typedef int (*coap_method_t)(struct coap_resource *resource,
+/**
+ * @typedef coap_observer_event_handler_t
+ * @brief Type of the handler being called when a resource's observers has been modified.
+ * Either an observer was added or removed.
```
missing docstring for the params

pdgendt commented on subsys/net/lib/coap/coap_server_shell.c:
```diff
@@ -0,0 +1,98 @@
+			    ++count,
+			    service->name,
+			    service->data->sock_fd < 0 ? "INACTIVE" : "ACTIVE",
+			    service->host,
```
This can be `NULL`, will add check.

pdgendt committed:
doc: connectivity: networking: api: Add CoAP server documentation

Created a dedicated page for CoAP service and resource examples.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

kartben reviewed (dismissed):
Very cool, thanks for all the work on this one! :)

jukkar commented on subsys/net/lib/coap/coap_server.c:
```diff
@@ -0,0 +1,762 @@
+	return coap_resource_remove_observer(resource, NULL, token, token_len);
+}
+
+static void coap_server_thread(void *p1, void *p2, void *p3)
```
Although this server lib is kind of doing similar things as what http server PR (#64465) is doing, it differs here. So we create a thread in library, and then it does the magic. In the http server lib, there is `http_server_init()` which loops through services that need to be created, and then opens the listening sockets. Then in `http_server_start()` it starts to wait client connections.
But it leaves the thread creation (which imho is kind of application related thing) to the application. Did you consider something like that here. Because this is a library, creating threads here is a bit problematic. I would leave as much of resource allocation (like threads and stacks) to the application side, as it is kind of messy configuration and mgmt wise.

@cfriedt you have been involved with the http server design from the beginning, what is you take on this?

pdgendt replied:
I did think about this, the main difference here, is that CoAP is UDP based and retransmissions are handled by the subsystem/application layer.
Having a dedicated thread for the application logic did simplify writing the server logic. But I do understand your concern, maybe @cfriedt can help indeed.

pdgendt replied:
As a side note, the CoAP client library also uses a separate thread in the subsystem.

cfriedt replied:
> Although this server lib is kind of doing similar things as what http server PR (#64465) is doing, it differs here.

It could be the case that #64465 (and the original GSoC code) still are not implementing the asynchronous server that was intended.

With an asynchronous server, there would not need to be any dedicated thread.

On the other hand, it is definitely simpler to use a thread and blocking I/O.

> `http_server_start()` it starts to wait client connections.

So that's where it went off course in a way. For an asynchronous server, we would have a kind of "tick" function for the server to call regularly (or on a per-event basis). Technically, a thread could be used even in that scenario, but all of would do is poll for events and then call the "tick" function.

I would suggest aiming for that kind of architecture in the long run, because technically, it is quite scalable and versatile. Of course, there will likelu be more statefulness. 

> @cfriedt you have been involved with the http server design from the beginning, what is you take on this?

I'm all for functional-first, so if the threaded server works, then that's ok. Optimization is a long process (sometimes too long).

Given that CoAP is (IIRC) a datagram-oriented protocol, I think it's significantly safer than HTTP in terms of resilience to DOS and timing attacks. Still have to take care for security and robustness of course.

jukkar commented on include/zephyr/net/coap.h:
```diff
@@ -841,8 +919,10 @@ bool coap_register_observer(struct coap_resource *resource,
+ * @return true if the observer was found and removed.
  */
-void coap_remove_observer(struct coap_resource *resource,
+bool coap_remove_observer(struct coap_resource *resource,
```
BTW, we are changing public APIs here, we probably need to put something to migration notes about these changes.

pdgendt replied:
Yes, is changing the return type from `void` to `bool` that intrusive? I guess its an issue if you have compiler warnings/error for unused return values.

jukkar replied:
Well, there is a compile error at least. It would just be great if any public API changes are documented somewhere and the migration doc seems to be a good place for those.

pdgendt replied:
I've added an item to the migration guide.

pdgendt committed:
doc: connectivity: networking: api: Add subsystem link notes to CoAP

The CoAP documentation has server/clients sample usages, added a
note about the available subsystems that can be used instead.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

pdgendt committed:
doc: releases: migration: Add CoAP API changes

Add a note to the 3.6 migration guide about the CoAP API changes
introduced with the CoAP server subsystem.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

cfriedt commented on samples/net/sockets/coap_server/src/main.c:
```diff
@@ -0,0 +1,85 @@
+
+#endif /* CONFIG_NET_IPV6 */
+
+COAP_SERVICE_DEFINE(coap_server, NULL, &coap_port, COAP_SERVICE_AUTOSTART);
```
Hmmm.... On one hand, this is clearly cloning some of the HTTP server concepts, but on the other hand, it seems to be implementing those concepts wrong. Of course, HTTP was not really pushed past the finish line by the GSoC contributor.

The "service define" concept was there to support one or more services defined by the application - not a single centralized service.

pdgendt replied:
It is possible to define multiple _services_ from the application, there is single _server_ thread which keeps track of these services.

pdgendt replied:
Perhaps the name `coap_server` for the service is a bit confusing 😄

cfriedt replied:
Sorry - my mistake - I was looking at this on mobile pre ☕ I didn't realize it was in the sample and thought it was part of the library 😅

Looks good
